core/templates/dev/head/pages/library/Library.js

```diff
@@ -80,6 +80,36 @@ oppia.controller('Library', [
       $http.get('/libraryindexhandler').success(function(data) {
         $scope.libraryGroups = data.activity_summary_dicts_by_category;
 
+        $scope.userIsLoggedIn = GLOBALS.userIsLoggedIn;
+        var activityDetails = {};
+        if ($scope.userIsLoggedIn) {
+          $scope.activitiesOwned = {explorations: {}, collections: {}};
+          $http.get('/creatordashboardhandler/data')
+            .then(function(response) {
+              $scope.libraryGroups.forEach(function(activityGroups) {
+                var activity = activityGroups.activity_summary_dicts;
+
+                activity.forEach(function(activity) {
+                  $scope.activitiesOwned.explorations[activity.id] = false;
+                  $scope.activitiesOwned.collections[activity.id] = false;
+
+                  response.data.explorations_list.forEach(
```

Comment:
Why the need for nested for loops?

Could we take this out of the current for loop and do a single pass through the list, updating $scope.activitiesOwned as needed?

Reply:
Actually what we have on ```$scope.libraryGroups``` is activities that are sorted according to categories like Humanities, Science etc. We can say it's a 2D array and the ```response.data.explorations_list``` is a 1D array. We'd require nested for loops for iterations. 
One thing that could be done is, storing all activities in one list and comparing it with the response we are getting. But that would again require one nested for loop.
forEach loops in angular do not support ```continue;``` or ```break;```, if we use for loops we could skip iterations once an activity is matched with the response and thereby reducing unnecessary iterations.

Reply:
There might be a misunderstanding here. I meant, why can't response.data.explorations_list.forEach(...) be taken outside the activity.forEach() loop? So you have the for loops in sequence, rather than nested -- giving you O(N) rather than O(N^2).

Reply:
I'm afraid I can't understand what you mean, each activity needs to be compared with the list of activities that are created by the current user. First loop is for iterating within the categories, second is for iterating within the activities inside the categories, third is for iterating and comparing each activity(present in library) with the ones published by the user. Currently the complexity is O(NxM) {where N is number of activities present on the dashboard and M is number of activities published}